fix: directory monitor input plugin when data format is CSV and csv_skip_rows>0 and csv_header_row_count>=1

etycomputer opened the pull request:
### Required for all PRs:

<!-- Complete the tasks in the following list. Change [ ] to [x] to
show completion. -->

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

<!-- Link to issues that describe the need for the change. Issues
should include context that will help reviewers understand why the
change is needed.

Make sure to link issues and using a keyword like "resolves #1234".
https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

resolves #9898
resolves #9903
resolves #5338

🐛 Fixed an issue with directory monitor input plugin when data format is CSV and csv_skip_rows>0 and csv_header_row_count>=1
✅ Added new test cases that cover these scenarios

etycomputer commented:
!signed-cla

etycomputer commented:
!signed-cla

etycomputer commented:
For this PR, did I have to create an issue first?

powersj commented:
It would be good to create an issue describing a bit more of what you are fixing, expected vs actual results, etc :) Then you can update the PR to reference that issue #.

Also, note that the lint tests failed, so you will want to clean those up.

Thanks!

etycomputer commented:
Hi @srebhan, can you please, let me know if you still think I need to make any changes.
Thank

etycomputer commented:
@srebhan, @powersj, and @Hipska

Hi,

I have made the requested changes to the CSV Parser.
The change is merging all the functionality between CSV.Parse() and CSV.ParseLine so that you can call both functions and they will perform the same action while keeping the state of the CSV parse updated between calls.
This allows the input plugins to call either of the functions and pass the while file or one row at a time.

Hope the fix is as expected.

Kind regards,
Ehsan

etycomputer commented:
Hi Joshua @powersj,
Could you have another look at this PR, please.
Cheers,
Ehsan

etycomputer commented:
> Yeah, it would need to be fixed in the parser, otherwise every plugin using this approach would need to be adapted.

Hi Thomas @Hipska,

I have made the requested changes in the csv parse and added tests to parser and both inputs.

Could you have a look and let me know if you have happy with the changes.

Thanks

etycomputer commented:
> Hey @etycomputer, thanks for the PR! I don't think this is the right place to fix this. Instead you should change the `CSV` parser's [expectation of getting all rows at once](https://github.com/influxdata/telegraf/blob/4c417976a34b25a17ce2b69e140fa5c5d8046621/plugins/parsers/csv/parser.go#L101) by decrementing the `SkipRows` by the rows you already skipped and check if there are remaining ones... This will then also fix `tail`.

Hi Sven @srebhan,

As requested I have made the changes to the csv parser and decremented the SkipRows and SkipHeaderCount to meet the requirements. I have also added multiple tests to cover the scenarios in the csv parser and both input plug-ins.

Hope the changes meet your requirements. Let me know if there is any other changes you need.

Kind regards,
Ehsan

srebhan commented on plugins/parsers/csv/parser_test.go:
```diff
@@ -332,17 +332,40 @@ func TestSkipRows(t *testing.T) {
 		},
 	)
-	testCSV := `garbage nonsense
+	testCSV1 := `garbage nonsense
```
Why is this necessary?

srebhan commented on plugins/parsers/csv/parser_test.go:
```diff
@@ -332,17 +332,40 @@ func TestSkipRows(t *testing.T) {
-	require.NoError(t, err)
-	require.Equal(t, "test_name2", metrics[0].Name())
-	require.Equal(t, expectedFields, metrics[0].Fields())
+	metrics0, err1 := p1.Parse([]byte(testCSV1))
+	require.NoError(t, err1)
+	require.Equal(t, "test_name2", metrics0[0].Name())
```
Please don't add random suffices. Keep tests as they are and add new ones. This way we can be sure that the code is still working as before.

srebhan commented on plugins/parsers/csv/parser_test.go:
```diff
@@ -375,15 +398,52 @@ func TestSkipColumnsWithHeader(t *testing.T) {
 	)
 	require.NoError(t, err)
 	testCSV := `col,col,col
-	1,2,3
-	trash,80,test_name`
+1,2,3
+trash,80,test_name`
```
Why is this necessary?

srebhan replied:
Can you please comment on why the reformatting is necessary?

srebhan replied:
I'm ok with removing them but they might be used to test the trim-space feature... If you don't mind I would feel better to keep them...

reimda replied:
Let's not remove the space. I don't see a reason to remove it.

etycomputer commented:
Hi @srebhan,

I am getting `typecheck  undeclared name: xxxx`.
I tried rebasing with Master, but that did not make a difference can you advise on this?

Kind regards,
Ehsan

etycomputer commented:
Hi @sspaink,

can you look at this job?
Made more changes and rebased with Master after your "chore: only check new issues with Go linter (#10054)" change.
This fixed one of the Lint checks now It is failing on another one.

Kind regards,
Ehsan

sspaink commented:
@etycomputer sorry for the problems with the linter, looks like this time it did find an error: https://github.com/influxdata/telegraf/pull/9865/files#diff-d6a8a85e74d13d871ed75dec8bd88f874b3e370f1b7df0262309c7eca5c237b7R74 the error returned by `NewParser` isn't being checked.

etycomputer commented:
> @etycomputer sorry for the problems with the linter, looks like this time it did find an error: https://github.com/influxdata/telegraf/pull/9865/files#diff-d6a8a85e74d13d871ed75dec8bd88f874b3e370f1b7df0262309c7eca5c237b7R74 the error returned by `NewParser` isn't being checked.

Thanks for pointing that out @sspaink. I clicked on the Details and could not figure out what the error was. I should have looked at the diff too. 

Cheers,
Ehsan

etycomputer commented:
Hey @sspaink, 

Sorry to bother you one more time, I am getting another failing test for a part of the code that I have not changed and it is only failing for Go 1.17.
Can you advise in this regard?

Thanks.

sspaink commented:
@etycomputer no worries, looks like it was a flaky test! I re-triggered the job and it passed, we will look into getting this test fixed.

srebhan commented on plugins/inputs/directory_monitor/directory_monitor.go:
```diff
@@ -285,24 +282,17 @@ func (monitor *DirectoryMonitor) parseFile(parser parsers.Parser, reader io.Read
+		m, err := parser.Parse(line)
+		if err == io.EOF {
+			return []telegraf.Metric{}, nil
 		}
-
-		m, err := parser.ParseLine(string(line))
 		if err != nil {
 			return nil, err
 		}
```
Please fuse the error checks and use `errors.Is` to also handle wrapped errors.
```golang
		if err != nil {
			if errors.Is(err, io.EOF) {
				return []telegraf.Metric{}, nil
			}
			return nil, err
		}
```
Same for tail.

srebhan commented on plugins/inputs/tail/tail.go:
```diff
@@ -378,13 +368,12 @@ func (t *Tail) receiver(parser parsers.Parser, tailer *tail.Tail) {
 		}
 
-		metrics, err := parseLine(parser, text, firstLine)
+		metrics, err := parseLine(parser, text)
```
You should also check for `EOF` here, shouldn't you?

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -96,22 +96,38 @@ func (p *Parser) compile(r io.Reader) *csv.Reader {
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	r := bytes.NewReader([]byte(line))
+	metrics, err := parseCSV(p, r)
+	if len(metrics) > 0 {
+		// only return the first metric as there should be only one
+		return metrics[0], err
+	}
+	return nil, err
```
Can you please change that to check there is only _exactly_ one metric?
```suggestion
	if err != nil {
		return nil, err
	}
	if len(metrics) != 1 {
		return nil, fmt.Errorf("expected 1 metric found %d", len(metrics))
	}
	return metrics[0], nil
```

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -96,24 +96,43 @@ func (p *Parser) compile(r io.Reader) *csv.Reader {
 	if !p.gotColumnNames {
-		headerNames := make([]string, 0)
-		for i := 0; i < p.HeaderRowCount; i++ {
+		for p.HeaderRowCount > 0 {
```
Any reason you don't fuse the `if` and `else` path as suggested above?

srebhan reviewed (commented):
Very nice update @etycomputer! I have some more comments and two comments left from the last review. The first one is the formatting of the test case. I try to understand if this is a necessary change. The second being folding the `else` case when there are column-names into the `if`. I think this will make the code more maintain and understandable. Is there a reason not to do it?

etycomputer commented:
Hi @srebhan, 

Thanks for looking at my code on such short notice.
I have applied the requested changes and added more explanation and my reasoning.

Regarding the last test format change (removed the spaces), I think the programmer that created the test initially did not really have the intention of adding the spaces. I think they thought it would look better and since it did not affect the result of the test it got missed. 
I think the spaces should be removed. 

Kind regards,
Ehsan

srebhan commented on plugins/inputs/directory_monitor/directory_monitor.go:
```diff
@@ -285,24 +282,17 @@ func (monitor *DirectoryMonitor) parseFile(parser parsers.Parser, reader io.Read
+		m, err := parser.Parse(line)
 		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return []telegraf.Metric{}, nil
```
Maybe doing this would be even better
```suggestion
				return nil, nil
```

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -96,49 +96,68 @@ func (p *Parser) compile(r io.Reader) *csv.Reader {
+	}
+	if len(metrics) == 1 {
+		return metrics[0], nil
+	} else if len(metrics) > 1 {
```
Sorry my bad, the linter will probably complain on this, just do
```suggestion
	}
	if len(metrics) > 1 {
```

srebhan reviewed (commented):
Nice, only some very minor things left. Please keep the spaces in the test-case as they also test for the 'trim-space' path I think.

etycomputer commented:
Hi @srebhan,

I have applied the minor changes.
Also, added a new test for the multi-metric error that we introduced.
Regarding the space removal, I strongly think that this should be removed. To make sure that the trim space functionality is still working as is as expected for multi header scenarios, I have added a new test. 

Hope that is acceptable to you. If you still think that adding the spaces is needed I am happy to add them back in.
I don't think the extra space is going to be a game-changer, and I really like my changes to get into master as soon as possible.
I would like to next work on a new feature related to the CVS parser.

Kind regards,
Ehsan

etycomputer commented:
How can we close issue #5338 
https://github.com/influxdata/telegraf/issues/5338
And issue #8989 
https://github.com/influxdata/telegraf/issues/9898
And issue #9903 
https://github.com/influxdata/telegraf/issues/9903
With this PR?

Hipska commented:
Add them in your description as you did for #9898. One per line.

More info: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

etycomputer commented:
Hey @Hipska,

How are you going?
Thanks, I will add them and try and see if I can find any other ones.

So, do you think this is ready to be merged or do we need to get feedback from anyone else like @srebhan?

Also, do I need to update the CHANGElOG or you is it done automatically?

Cheers,
Ehsan

srebhan reviewed (approved):
LGTM. That's an amazing PR @etycomputer. I like the test solution adding another test. Thank you very much for your work!

srebhan commented:
> So, do you think this is ready to be merged or do we need to get feedback from anyone else like @srebhan?

After my approval (which you just got) I set the "ready for final review" tag. With this someone from the InfluxData telegraf team will give the PR another review and, if he/she does not have any comment, merge it.

> Also, do I need to update the CHANGELOG or you is it done automatically?

You don't have to care for this, it's done (semi)automatically IIRC.

etycomputer commented:
Thank you very much @srebhan you were also have me good comments and pointed out excellent alternative solution. I really enjoyed working with you on this PR.

Kind regarding,
Ehsan

etycomputer commented:
Hey @powersj, 

Are you able to do the final review and push my changes into master, please?

Cheers,
Ehsan

srebhan commented:
@etycomputer please be patient waiting for the final review. Rest assured that your PR is on the list and will not be forgotten. As the change is touching a tricky part of telegraf, it might be possible your PR will only be included in the next version (1.21).

etycomputer commented:
Hey @srebhan,

Sorry, the reason I wanted this change to go in was because my metadata feature that I wanted to add depends on these changes.

I have already based my new PR #10083 off this branch. So most probably they will all go in at the same time.

Is there a timeline when you are expecting to go to the next version?

It would be awesome if you could also have a look at #10083 and let me know what you think.

I have just started programming in Go and I really like it so far.

I would be happy to work on some of the smaller jobs if you come across any please let me know.

Kind regards,
Ehsan

reimda commented on plugins/parsers/csv/parser_test.go:
```diff
@@ -343,6 +385,30 @@ hello,80,test_name2`
+	m, err = p.ParseLine(testCSVRows[2])
+	require.NoError(t, err)
+	require.Equal(t, "test_name2", m.Name())
+	require.Equal(t, expectedFields, m.Fields())
```
We need to check that the tag "line1=hello" is in the metric.

reimda commented on plugins/parsers/csv/parser_test.go:
```diff
@@ -59,9 +60,33 @@ func TestHeaderOverride(t *testing.T) {
+		},
+	)
+	require.NoError(t, err)
+	metrics, err = p.Parse([]byte(testCSVRows[0]))
```
Do you mean to call ParseLine instead of Parse here?

etycomputer commented:
Hey @reimda,

Thanks for reviewing my work.
I have added the extra test you requested.

Regarding your other comment, I have chosen to use the combination to cover scenarios that someone may use the parse.Parse() for its first line and the parse.ParseLine() for any line after that. This was the approach that is currently is being used on the master branch but going forward I have removed the need to differentiate between the initial line (header column) and other lines (records).

Let me know if you like to change the p.parse() to p.ParseLine().

Kind regards,
Ehsan

srebhan commented:
> I would be happy to work on some of the smaller jobs if you come across any please let me know.

@etycomputer, you might want to go through [`help wanted` PRs](https://github.com/influxdata/telegraf/pulls?q=is%3Aopen+is%3Apr+label%3A%22help+wanted%22) or check the issues (best would be [bug reports](https://github.com/influxdata/telegraf/issues?q=is%3Aopen+is%3Aissue+label%3Abug)) if you can find anything you like. Furthermore, you might want to join our Slack channel #telegraf-dev and/or #telegraf. Happy to see you there. :-)

etycomputer commented:
Thank you All for helping me with this PR.

srebhan commented:
@etycomputer thank you for fixing this!
